### ISSUE TITLE
New option in "object-curly-spacing" for empty objects

### ISSUE THREAD

**Issue by feross** [eslint/eslint#7927]
**Tell us about your environment**

* **ESLint Version:** v3.8.0
* **Node Version:** v6.7.0
* **npm Version:** v3.10.7

**What parser (default, Babel-ESLint, etc.) are you using?**

Default

**Please show your full configuration:**

```
{
  "rules": {
    "object-curly-spacing": ["error", "never"]
  }
}
```

**What did you do? Please include the actual source code causing the issue.**

```js
var a = { }
```

**What did you expect to happen?**

I expected that the rules should complain about the stray space between the curly braces { }

I also think it would be a good idea to add a separate config for empty objects, since I think that people who have this rule set to "always" still doesn't want any whitespace inside empty objects.

**What actually happened? Please include the actual, raw output from ESLint.**

No error was raised.

(Note: Re-opening https://github.com/eslint/eslint/issues/7380 for discussion, as suggested)

**Comment by ilyavolodin** [comment]
I think this should be treated as a bug. I was able to reproduce this issue, and I think this is not an intended behavior, since the configuration clearly states that I don't want any spaces between curly braces.

**Comment by not-an-aardvark** [comment]
How should the rule handle this case?

```js
/* eslint object-curly-spacing: [error, "always"] */

var foo = {};

var bar = { };
```

I don't think the first object should be an error -- the object is technically missing a space, but I'm pretty sure `{}` is the generally preferred style even in codebases that use `object-curly-spacing: always`.

**Comment by ilyavolodin** [comment]
@not-an-aardvark I think it should be reported as an error. We can create another option for this rule, something like `always-not-empty` (I know, terrible name), that would would not report empty objects, but I think with the `always` option it should be reported.

**Comment by ljharb** [comment]
Does the "never" option report on `bar` above, or does it have the same problem?

**Comment by not-an-aardvark** [comment]
@ljharb Currently, empty objects are completely ignored (see the original issue).

@ilyavolodin I'm fine with adding a new option, but I'm concerned about changing the default behavior of the `always` option. I suspect a lot of projects currently have `{}` with `always` and prefer it that way.

**Comment by ilyavolodin** [comment]
@not-an-aardvark Hmm... It's a fair point, but on the other hand it feels a bit strange that the option is called `always`, but yet doesn't flag cases when there are no spaces... But I get what you are saying. Let's see what others think.

**Comment by platinumazure** [comment]
Can we change the default behavior (and maybe the rule default, if it's "always", to "always-except-empty" or whatever we decide on) in a major release? That way users would already be aware of possibly having to migrate anyway and we can get the rule to be sensible/consistent.

**Comment by ljharb** [comment]
Adding an option now in a minor, and changing the default later in a major, would also enable people to be robust against the breaking change later by explicitly specifying the option.

**Comment by pmcelhaney** [comment]
I don't care about the difference between `{ }` and `{}` but I work with someone who refuses to let it go and continues to clog up pull requests with bike shed comments.

I'm going to work on a PR to make the treatment of empty objects an option.

- `spaceInEmptyObject: "always"`: Require a space inside of an empty object literal (`{ }` is valid)
- `spaceInEmptyObject: "never"`: Disallow a space inside of an empty object literal (`{}` is valid)

If `spaceInEmptyObject` is not specified, either `{ }` or `{}` is allowed.

If you don't hear from me within the next week, assume that I got too busy and gave up on the idea.

**Comment by pmcelhaney** [comment]
Thanks for the feedback on the pull request. I think I understand how to satisfy the following criteria at the same time:

1. Provide a work-around for the reported bug that's not a breaking change
2. Make the API more flexible
3. Make the API easier to understand
4. Establish a clear path to make `"object-curly-spacing": ["error", "never"]` more consistent in an upcoming major release (i.e. fix this bug)

I'd like to proceed by reopening #7380 and implementing it in three phases:

1. Update the documentation and build consensus around that.
2. Add tests so we can make sure we're on the same page and we've covered every edge case.
3. Write the code to make the tests pass, get it reviewed, and merge.

I'll move forward if I get two thumbs up from contributors on this comment.

**Comment by pmcelhaney** [comment]
Thanks @nzakas. I'm no longer interested in completing this enhancement. If someone wants to pick it up you're welcome to use the work from my PR.

**Comment by nzakas** [comment]
Thanks for the response, we will see if anyone else wants to pick it up.

**Comment by not-an-aardvark** [comment]
I'm going to go ahead and mark this as "evaluating" again -- the `accepted` label was originally added when this was considered a bug, but now it seems to be prosing a new option.

**Comment by nzakas** [comment]
It seems like there is confusion over the purpose of this issue and if it has changed from the original request. As such, I think the best path forward is for any interested parties to open a new issue with a proposal that we can evaluate. Closing this issue now, but anyone interested should feel free to open a new issue and just point back to this one as the origin.
